From 0dd6f599d7355910a96a5983f929dfc577d05ded Mon Sep 17 00:00:00 2001
From: Wu Zhangben <wuzhangben@eswincomputing.com>
Date: Wed, 13 Aug 2025 14:46:30 +0800
Subject: [PATCH 478/534] feat:lpcpu config distribute by linux driver

Changelogs:

1. Use eswin-lpcpu driver to distribute configs to lpcpu
2. Rewrite eswin-lpcpu driver to support reentrant
3. Load npu firmware which for lpcpu warmup to reserved memory by lpcpu driver
4. Update dts file for 2-die boards

Signed-off-by: Wu Zhangben <wuzhangben@eswincomputing.com>
---
 .../eic7702-deepcomputing-fml13v03-evt.dts    | 15 ++++++----
 .../dts/eswin/eic7702-som560-interleave.dtsi  | 29 ++++++++++---------
 2 files changed, 26 insertions(+), 18 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
index fce5c14cea74..ea18664d8de9 100755
--- a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
@@ -119,7 +119,7 @@ ramoops@fff00000 {
 			no-map;
 		};
 
-		g2d_8GB_boundary_reserved_4k { // 32M for npu0 runtime
+		lpcpu_reserved: g2d_8GB_boundary_reserved_4k { // 32M reserved for lpcpu
 			no-map;
 			reg = <0x1 0xfe000000 0x0 0x2000000>;
 		};
@@ -145,7 +145,7 @@ d1_g2d_4GB_boundary_reserved_4k {
 			reg = <0x20 0xfffff000 0x0 0x1000>;
 		};
 
-		d1_g2d_8GB_boundary_reserved_4k { // 32M for npu1 runtime
+		d1_lpcpu_reserved: d1_g2d_8GB_boundary_reserved_4k { // 32M reserved for lpcpu
 			no-map;
 			reg = <0x21 0xfe000000 0x0 0x2000000>;
 		};
@@ -504,14 +504,19 @@ &d1_ipc_scpu {
 
 &d0_lpcpu {
 	status = "okay";
+	memory-region = <&lpcpu_reserved>;
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_die0_gpio0_default &pinctrl_die1_gpio7_on_default>;
-	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>, <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
+	pinctrl-0 = <&pinctrl_die0_gpio0_default>;
+	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>;
 	vddctrl = <104 0 1>;
 };
 
 &d1_lpcpu {
-	status = "disabled";
+	status = "okay";
+	memory-region = <&d1_lpcpu_reserved>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die1_gpio7_on_default>;
+	irq-gpios = <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
 };
 
 &pinctrl_die0_gpio25_default {
diff --git a/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi b/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
index cf97e2daabab..893401ea8268 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
+++ b/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
@@ -168,10 +168,10 @@ d1_dc_reserved: d1_dc_reserved@2260000000 {
 		// 	reg = <0x0 0xfffff000 0x0 0x1000>;
 		// };
 
-		// g2d_8GB_boundary_reserved_4k {
-		// 	no-map;
-		// 	reg = <0x1 0xfffff000 0x0 0x1000>;
-		// };
+		lpcpu_reserved: g2d_8GB_boundary_reserved_4k { // 32M reserved for lpcpu
+			no-map;
+			reg = <0x1 0xfe000000 0x0 0x2000000>;
+		};
 
 		// g2d_12GB_boundary_reserved_4k {
 		// 	no-map;
@@ -187,12 +187,10 @@ nid_1_zero_device_simu: nid_1_zero_device_simu@2040000000 {
 		// 	reg = <0x20 0xfffff000 0x0 0x1000>;
 		// };
 
-		// d1_g2d_8GB_boundary_reserved_4k {
-		// 	no-map;
-		// 	reg = <0x21 0xfffff000 0x0 0x1000>;
-		// };
-
-
+		d1_lpcpu_reserved: d1_g2d_8GB_boundary_reserved_4k { // 32M reserved for lpcpu
+			no-map;
+			reg = <0x21 0xfe000000 0x0 0x2000000>;
+		};
 	};
 
 	distance-map {
@@ -448,14 +446,19 @@ &d1_ipc_scpu {
 
 &d0_lpcpu {
 	status = "okay";
+	memory-region = <&lpcpu_reserved>;
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_die0_gpio0_default &pinctrl_die1_gpio7_on_default>;
-	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>, <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
+	pinctrl-0 = <&pinctrl_die0_gpio0_default>;
+	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>;
 	soc-vdd-ctrl = <21 0 0>;
 };
 
 &d1_lpcpu {
-	status = "disabled";
+	status = "okay";
+	memory-region = <&d1_lpcpu_reserved>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_die1_gpio7_on_default>;
+	irq-gpios = <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
 };
 
 &pcie {
-- 
2.53.0

